src/types.ts

```diff
@@ -0,0 +1,13 @@
+import { RouterContext } from '@koa/router';
+import Koa, { DefaultState, Middleware } from 'koa';
+import Articles from './articles';
+
+export type AppState = DefaultState;
+
+export type AppContext = RouterContext<AppState, {
+  articles: Articles;
+}>;
+
+export type AppMiddleware = Middleware<AppState, AppContext>;
+
+export type App = Koa<AppState, AppContext>;
```

Comment:
Did want these inside `app.ts`, but end up with circular references otherwise.

Reply:
Is `App` required? It should be inferred correctly from the default export returning in `app.ts`

Reply:
Looking at whether it can be reduced to just `AppContext`, as that's the important one.

Reply:
Nope. The separate router is the problem...

Reply:
In fact, even moving `createRouter` doesn't help (since the routes themselves are now circular).

Reply:
Might have to inject the router into the app factory? Shame, as the `apiDocumentation` middleware relies on a specific route. Might have to inject that too?

(Feels like testing will be a bit weirder.)

Reply:
You could avoid that by making the route generic? 

```typescript
function createRouter<State, Context>() {
  const router = new Router<State, Context>();

  router.get( 
     //... 
  );
}
```
Although might need to narrow the types on the generics

Reply:
Tried that, not assignable in https://github.com/libero/article-store/blob/1acc2dadafdd154f9b6f5f61fb9545524da29be3/src/app.ts#L17 😞

Reply:
Thinking about making the path an argument on `apiDocumentation`, so it's decoupled from the router.

Reply:
https://github.com/libero/article-store/pull/60/commits/bf78b85d521ce96af83994c6611e7a783b795a60, but needs extracting

Reply:
Right, I see. Pulled it down to check and yeh it's like the app and router are pretty must one and the same conceptually. I definitely see the benefit in splitting out the routes though.

In this case though, it does make sense to have some types grouped together as a common concept that they are all a part of (the "app").

Reply:
Push https://github.com/libero/article-store/pull/60/commits/14962894eb80dce81cceb5f12cba8fc23c8a479e which is now possible. Doesn't feel great though.